nidx/nidx_paragraph/src/resource_indexer.rs

```diff
@@ -79,9 +80,21 @@ pub fn index_paragraphs(
             let end_pos = p.end as u64;
             let index = p.index;
             let split = &p.split;
-            let lower_bound = std::cmp::min(start_pos as usize, chars.len());
-            let upper_bound = std::cmp::min(end_pos as usize, chars.len());
-            let text: String = chars[lower_bound..upper_bound].iter().collect();
+
+            // The text of the paragraph can be specified in the paragraph message, otherwise slice it from the full field text.
+            let text_ref: String;
+            if let Some(t) = &p.text {
+                text_ref = t.to_string();
+            } else {
+                let lower_bound = std::cmp::min(start_pos as usize, chars.len());
+                let upper_bound = std::cmp::min(end_pos as usize, chars.len());
+                if lower_bound >= upper_bound {
+                    warn!("Skipping paragraph with wrong start and end positions");
+                    continue;
+                }
+                let slice = &chars[lower_bound..upper_bound];
+                text_ref = slice.iter().collect();
```

Comment:
With the `min(start_pos, chars.len)` thing we still could ignore invalid ranges. Maybe move it to a check in the if, for example:

```suggestion
                if start_pos >= end_pos || end_pos > chars.len() {
                    warn!("Skipping paragraph with wrong start and end positions");
                    continue;
                }
                let slice = chars[start_pos..end_pos];
                text_ref = slice.iter().collect();
```